lua/paq.lua

```diff
@@ -226,95 +187,365 @@ local function log_update_changes(pkg, prev_hash, cur_hash)
     end)
 end
 
-local function pull(pkg, counter, sync)
-    local prev_hash = lock[pkg.name] and lock[pkg.name].hash or pkg.hash
-    call_proc("git", { "pull", "--recurse-submodules", "--update-shallow" }, pkg.dir, function(ok)
+---@param name string
+---@param msg_op Messages
+---@param result boolean
+---@param n integer
+---@param total integer
+local function report(name, msg_op, result, n, total)
+    local count = n and string.format(" [%d/%d]", n, total) or ""
+    vim.notify(
+        string.format(" Paq:%s %s %s", count, msg_op[result], name),
+        result == "err" and vim.log.levels.ERROR or vim.log.levels.INFO
+    )
+end
+
+---Object to track result of operations (installs, updates, etc.)
+---@param total integer
+---@param callback function
+local function new_counter(total, callback)
+    return coroutine.wrap(function()
+        local c = { ok = 0, err = 0, nop = 0 }
+        while c.ok + c.err + c.nop < total do
+            local name, msg_op, result = coroutine.yield(true)
+            c[result] = c[result] + 1
+            if result ~= "nop" or Config.verbose then
+                report(name, msg_op, result, c.ok + c.nop, total)
+            end
+        end
+        callback(c.ok, c.err, c.nop)
+        return true
+    end)
+end
+
+
+-- }}}
+-- LOCKFILE: {{{
+
+local function lock_write()
+    -- remove run key since can have a function in it, and
+    -- json.encode doesn't support functions
+    local pkgs = vim.deepcopy(Packages)
+    for p, _ in pairs(pkgs) do
+        pkgs[p].build = nil
+    end
+    local file = uv.fs_open(Config.lock, "w", 438)
+    if file then
+        local ok, result = pcall(vim.json.encode, pkgs)
+        if not ok then
+            error(result)
+        end
+        assert(uv.fs_write(file, result))
+        assert(uv.fs_close(file))
+    end
+    Lock = Packages
```

Comment:
Should this be a `deepcopy`? In case the user calls `paq.__call` again.

Reply:
Previously I was using `deepcopy` but it doesn't really make a difference. I opted to this solution since is cheaper to pass the table by reference instead of coping deeply the entire content. But we could use either and it will work the same.